### PR TITLE
Add /blocks , /blocks/{hashOrHeight} APIs

### LINKED ISSUE #407: API: Block
Client API for Blocks.
It may be required by exchanges and explorer.

/blocks
/blocks/:hash
/blocks/:hash/transactions
/blocks/:hash/operations

### PR DESCRIPTION
**Pull request by anarcher**
### Github Issue

Fixes #407 

### Background

- `/api/blocks`
- `/api/blocks/{hashOrHeight}`

### Solution

-  [x] Add block handlers (`/api/blocks` , `/api/blocks/{hash}` )
-  [x] Use cursor as height
-  [x] Fix paging issue ㅠㅠ 
-  [x] Test codes
-  [x] Documention

### CODE REVIEW

**Comment by anarcher** [comment]
```
$http localhost:2821/api/v1/blocks/219
HTTP/1.1 200 OK
Content-Length: 618
Content-Type: application/hal+json
Date: Thu, 15 Nov 2018 08:16:17 GMT
X-Ratelimit-Limit:
X-Ratelimit-Remaining:
X-Ratelimit-Reset:

{
    "_links": {
        "self": {
            "href": "/api/v1/blocks/4WML58QcH7UxSTeg5UrDfooeyRqubqH8gTM5B6BN8Wkf"
        }
    },
    "confirmed": "2018-11-15T16:45:07.789370000+09:00",
    "hash": "4WML58QcH7UxSTeg5UrDfooeyRqubqH8gTM5B6BN8Wkf",
    "height": 219,
    "prev_block_hash": "vp4PqRdVmHwi5rSrcxNe2HDDwUKzgrKFKnZWtDLDwLZ",
    "proposed_time": "2018-11-15T16:45:07.776943000+09:00",
    "proposer": "GDIRF4UWPACXPPI4GW7CMTACTCNDIKJEHZK44RITZB4TD3YUM6CCVNGJ",
    "proposer_transaction": "BtZVJMHCj7EMAvujSnQero52Mrc9jyQRnoJ34V2WZdKG",
    "round": 0,
    "transactions": [
        "FLWD4L9DKfEfh8bDCfJz7qBZFPc17tRPg3fVTNeQhjRq"
    ],
    "transactions_root": "CzvSaTF4aLopfwopj2AssFy9PJRekEkxWWqvvadHgouQ",
    "version": 0
}
```

```
$http localhost:2821/api/v1/blocks/4WML58QcH7UxSTeg5UrDfooeyRqubqH8gTM5B6BN8Wkf
HTTP/1.1 200 OK
Content-Length: 618
Content-Type: application/hal+json
Date: Thu, 15 Nov 2018 08:16:40 GMT
X-Ratelimit-Limit:
X-Ratelimit-Remaining:
X-Ratelimit-Reset:

{
    "_links": {
        "self": {
            "href": "/api/v1/blocks/4WML58QcH7UxSTeg5UrDfooeyRqubqH8gTM5B6BN8Wkf"
        }
    },
    "confirmed": "2018-11-15T16:45:07.789370000+09:00",
    "hash": "4WML58QcH7UxSTeg5UrDfooeyRqubqH8gTM5B6BN8Wkf",
    "height": 219,
    "prev_block_hash": "vp4PqRdVmHwi5rSrcxNe2HDDwUKzgrKFKnZWtDLDwLZ",
    "proposed_time": "2018-11-15T16:45:07.776943000+09:00",
    "proposer": "GDIRF4UWPACXPPI4GW7CMTACTCNDIKJEHZK44RITZB4TD3YUM6CCVNGJ",
    "proposer_transaction": "BtZVJMHCj7EMAvujSnQero52Mrc9jyQRnoJ34V2WZdKG",
    "round": 0,
    "transactions": [
        "FLWD4L9DKfEfh8bDCfJz7qBZFPc17tRPg3fVTNeQhjRq"
    ],
    "transactions_root": "CzvSaTF4aLopfwopj2AssFy9PJRekEkxWWqvvadHgouQ",
    "version": 0
}
```

**Comment by anarcher** [comment]
`/blocks` 's cursor is height. 

```
$http localhost:2821/api/v1/blocks cursor==209 limit==10 | jq ._embedded.records[].height
209
210
211
212
213
214
215
216
217
218
```

```
$http localhost:2821/api/v1/blocks cursor==209 limit==10 | jq ._links
{
  "next": {
    "href": "/api/v1/blocks?cursor=218&limit=10&reverse=false"
  },
  "prev": {
    "href": "/api/v1/blocks?cursor=218&limit=10&reverse=true"
  },
  "self": {
    "href": "/api/v1/blocks?cursor=209&limit=10"
  }
}
```

**Comment by Charleslee522** [comment]
The `integration_tests` will pass after #735

**Comment by spikeekips** [comment]
@anarcher Your checklist is getting bigger :)

**Comment by Charleslee522** [comment]
@spikeekips Yes~~~ I'll make it bigger but stable :)

**Comment by anarcher** [comment]
I finished the APIs `/blocks` and `/blocks/{hash}` and docs for it. 

If this apis `/blocks/{hash}/transacions` and `/blocks/{hash}/operations` are needed, I think this apis will be made  with other PRs.  :->

**Comment by spikeekips** [comment]
@anarcher Magna!!

**Comment by anarcher** [comment]
plz, review it. @spikeekips @cmcm2222 @Charleslee522 @kfangw

File: lib/node/runner/api/blocks.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+	var (
+		cursor []byte // cursor as height
+		blocks []resource.Resource
+	)
+
+	var option *storage.WalkOption
+	{
+		height, err := strconv.ParseUint(string(p.Cursor()), 10, 64)
+		if err != nil {
+			height = 1 // default cursor is height 1
```
Comment (spikeekips) [inline lib/node/runner/api/blocks.go]: We already defined, `common.GenesisBlockHeight`

File: lib/storage/leveldb.go [hunk 2]
```diff
@@ -402,6 +405,10 @@ func (st *LevelDBBackend) Walk(prefix string, option *WalkOption, walkFunc WalkF
 
 	var cnt uint64 = 0
 
+	if option.SkipCursor == true {
+		iter.Seek(st.makeKey(cursor))
```
Comment (spikeekips) [inline lib/storage/leveldb.go]: If `SkipCursor` is `true`, the item of cursor should be skipped, right?  If `SkipCursor` is `false`, the item of cursor will not be included to `walkFunc`. Before calling `iter.Next()`(`iterFunc`), this should be call `iter.Key()` and `iter.Value()` for cursor item.

Comment (anarcher) [reply lib/storage/leveldb.go]: If `SkipCursor` is true,  the item is cursor skip by `iter.Seek`. 
if `SkipCursor` is false, the item is cursor is included in iterator.

File: lib/node/runner/api/blocks.go (hunk 1 above)
Comment (anarcher) [reply lib/node/runner/api/blocks.go]: I changed it :->

File: lib/storage/leveldb.go (hunk 2 above)
Comment (spikeekips) [reply lib/storage/leveldb.go]: I think `iter.Seek()` does not skip the cursor item. Please correct me :)

Comment (anarcher) [reply lib/storage/leveldb.go]: Oh, Thanks. @spikeekips :-)

File: lib/node/runner/api/block_test.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+	"encoding/json"
+	"io/ioutil"
+	"strings"
+	"testing"
+
+	"boscoin.io/sebak/lib/block"
+	"github.com/stretchr/testify/require"
+)
+
+func TestBlockHAndler(t *testing.T) {
```
Comment (kfangw) [inline lib/node/runner/api/block_test.go]: TestBlockHAndler -->
TestBlockHandler 
:)

**Pull request merged**